File: packages/mirascript/src/vm/operations/type-check.ts
import { VmError } from '../../helpers/error.js';
import { getVmType, isVmArray, isVmRecord } from '../../helpers/types.js';
import type { VmTypeName, VmAny, VmArray, VmRecord, VmValue } from '../types/index.js';
import { $AssertInit } from './common.js';

/** 获取值的类型名称 */
export const $Type = (value: VmAny): VmTypeName => {
    // 允许未初始化值通过
    if (value == null) return 'nil';
    return getVmType(value);
};
/** 判断值是否为布尔值 */
export const $IsBoolean: (value: VmAny) => value is boolean = (value) => {
    $AssertInit(value);
    return typeof value == 'boolean';
};
/** 判断值是否为数字 */
export const $IsNumber: (value: VmAny) => value is number = (value) => {
    $AssertInit(value);
    return typeof value == 'number';
};
/** 判断值是否为字符串 */
export const $IsString: (value: VmAny) => value is string = (value) => {
    $AssertInit(value);
    return typeof value == 'string';
};
/** 判断值是否为记录 */
export const $IsRecord: (value: VmAny) => value is VmRecord = (value) => {
    $AssertInit(value);
    return isVmRecord(value);
};
/** 判断值是否为数组 */
export const $IsArray: (value: VmAny) => value is VmArray = (value) => {
    $AssertInit(value);
    return isVmArray(value);
};
/** 断言值非 nil */
export const $AssertNonNil: <T extends VmValue>(value: T | undefined) => asserts value is NonNullable<T> = (value) => {
    $AssertInit(value);
    if (value !== null) return;
    throw new VmError(`Expected non-nil value`, null);
};

File: crates/python/mirascript/_vm/operations/type_check.py
from __future__ import annotations

from ..._helpers.constants import Uninitialized
from ..._helpers.types import get_vm_type, is_vm_array, is_vm_record
from ..._helpers.checker import is_number
from ..error import VmError
from ..types import VmAny
from .common import AssertInit


def Type(val: VmAny):
    if val is Uninitialized or val is None:
        return "nil"
    return get_vm_type(val)


def IsBoolean(val: VmAny) -> bool:
    AssertInit(val)
    return isinstance(val, bool)


def IsNumber(val: VmAny) -> bool:
    AssertInit(val)
    return is_number(val)


def IsString(val: VmAny) -> bool:
    AssertInit(val)
    return isinstance(val, str)


def IsRecord(val: VmAny) -> bool:
    AssertInit(val)
    return is_vm_record(val)


def IsArray(val: VmAny) -> bool:
    AssertInit(val)
    return is_vm_array(val)


def AssertNonNil(val: VmAny) -> None:
    AssertInit(val)
    if val is not None:
        return
    raise VmError("Expected non-nil value", None)
